### ISSUE TITLE
Create Indexer for Sputnik DAO

### ISSUE THREAD

**Issue by ori-near** [NEAR-DevHub/neardevhub-treasury-dashboard#157]
### Description
We need to build an indexer similar to the one used for [DevHub proposals](https://github.com/NEAR-DevHub/neardevhub-api-rs), designed to index proposals from Sputnik DAOs. This will enable enhanced functionality for filtering and searching within the Treasury Dashboard.

#### Decide Indexing Strategy
- Should we index all .sputnik-dao.near accounts, or only the accounts associated with our treasury?
- If we choose to index treasury-related accounts only, we need to implement an API to dynamically start indexing new accounts created through the self-create flow.

#### Data to be Indexed
- Proposal data as retrieved from the DAO contract.
- Execution transaction details, including the link and timestamp.

##### Sorting
- By creation time.
- By expiry time.

##### Filtering
- By creator.
- By proposal type.
- By number of votes.
- By proposal status.

##### Searching
- By proposal description.

**Comment by ori-near** [comment]
Notes from our standup call: Ideally create one indexer for all DAOs, leverage sharding ... instead of deploying an indexer per DAO instance.

**Comment by Tguntenaar** [comment]
# Game Plan for Sputnik DAO Indexer

---

## Indexing Strategy

We will use a centralized indexer to handle all DAOs rather than deploying a separate instance for each DAO. The endpoints will support filtering and querying by DAO account.

Since our Fly.io machines remain offline unless needed, we’ll implement a "just-in-time" indexing strategy. This means we’ll only index data when the frontend requests it for a specific DAO account. This approach will index all relevant data, but only when an updated table for a DAO account is requested.

To fetch uncached transactions, we’ll index each DAO subaccount separately using the Nearblocks API. This ensures we retrieve all transactions not already stored in our cache.

Given the relatively small volume of data, all transactions can be stored in a dedicated table. From there, we’ll transform this raw data into specific tables optimized for sorting, filtering, and searching.

Additionally, we can index the DAO factory to detect and dynamically start indexing new accounts created through the self-create flow.

---

## Data to Be Indexed

Most of the proposal data is already indexed, including searching by description, which is functional. However, we need to index the following new data:  

- **Execution transaction details**, including:
  - Transaction link (e.g., to Explorer).
  - Execution timestamp.

---

## New API Endpoint

The new endpoint will allow querying transactions with sorting and filtering options:

- **`/txns/{DAO_ID}/?sort={sorting}&filters={filters}`**  
  - **Sorting Options**:
    - By creation time.
    - By expiry time.
  - **Filters**:
    - By creator.
    - By proposal type.
    - By number of votes.
    - By proposal status.

This endpoint will integrate seamlessly with the frontend, providing the necessary data for the Treasury Dashboard.

**Comment by Tguntenaar** [comment]
Note that sorting by creation time and expiry time will be the same since the expiry time is defined by the creation time + policy duration time.

**Comment by Tguntenaar** [comment]
## Update

I'm currently working on indexing `add_proposal` and `act_proposal` functions from the DAO contract. 

At first I wanted to use the transaction arguments to indext the state. However these functions change the proposal's inside the contract with a ton of logic. 

So instead of recreating that we just get the proposal from contract on the transaction blockheight. I'm now (Friday Dec 20) working on deserializing the string U128. So we can use the RPC response to cache the proposal.

@frol do you have a tip to deserialize this?

```
JsonRpcError(ServerError(HandlerError(ContractExecutionError { vm_error: "wasm execution failed with error: HostError(GuestPanic { panic_msg: \"panicked at 'ERR_NO_PROPOSAL', sputnikdao2/src/views.rs:102:48\" })", block_height: 114574681, block_hash: CEx1UdXoBRxsBFoHys9m7zaMwQsRP9e7qiVvHZWW1erf })))
DeserializeError(Error("u128 is not supported", line: 1, column: 2807))
```

**Comment by Tguntenaar** [comment]
# Update

The last deserialisation error has been fixed.

Theses fields are indexed properly:
- id
- description
- kind
- proposer
- status
- hash

I'm currently working on indexing the following fields:
- [ ] tx_timestamp
- [ ] vote_counts
- [ ] votes
- [ ] submission_time
- [ ] proposal_action

**Comment by ori-near** [comment]
Note, this ticket has a dependency on https://github.com/near-daos/sputnik-dao-contract/issues/214, which @Tguntenaar will work on.

**Comment by Tguntenaar** [comment]
### Update

I'm currently writing tests with `near-workspace-rs` that deploys the DAO contract to make sure it get's indexed correctly. We still use the nearblocks API to get the transaction hash for issue [92](https://github.com/NEAR-DevHub/neardevhub-treasury-dashboard/issues/92).  

After that we can finish the front end PR which is scheduled for tomorrow.

**Comment by Tguntenaar** [comment]
## Update

The tests on indexing based on changelog instead of nearblocks keep failing in this [PR](https://github.com/NEAR-DevHub/neardevhub-api-rs/pull/18/). Once that is merged those changes need to be merged into the sputnik DAO indexer in this PR [11](https://github.com/NEAR-DevHub/neardevhub-api-rs/pull/11). In order to test the dao indexer.

I've made another update for the front end PR [here](https://github.com/NEAR-DevHub/neardevhub-treasury-dashboard/issues/70).

**Comment by ori-near** [comment]
@frol please re-assign this when ready. It is blocking various features, including the export.

**Comment by ori-near** [comment]
@Norfolks

**Comment by Norfolks** [comment]
Hey

**Comment by Megha-Dev-19** [comment]
Indexer is [built](https://github.com/near-daos/sputnik-dao-caching-api-server/tree/main) and deployed. 
`https://sputnik-indexer-divine-fog-3863.fly.dev`

**Comment by Norfolks** [comment]
@Megha-Dev-19 sure, I will make a README today or tommorow.
